File: Python/501. Find Mode in Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    def __init__(self):
        self.ans = list()
        self.cnt = 0
        self.tmp = dict()
        
    def findMode(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return []
        def dfs(node):
            if node is None:
                return None
            self.tmp[node.val] = self.tmp.get(node.val, 0) + 1
            if self.tmp[node.val] > self.cnt:
                self.cnt = self.tmp[node.val]
                self.ans = [node.val]
            elif self.tmp[node.val] == self.cnt:
                self.ans.append(node.val)
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        return self.ans

File: C++/501. Find Mode in Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> ans;
    int cnt;
    unordered_map<int, int>maple;
    void dfs(TreeNode* node)
    {
        if(node==nullptr)
            return ;
        maple[node->val]++;
        if(maple[node->val] > cnt)
        {
            cnt = maple[node->val];
            ans.clear();
            ans.push_back(node->val);
        }
        else if(maple[node->val] == cnt)
            ans.push_back(node->val);
        dfs(node->left);
        dfs(node->right);
    }
    vector<int> findMode(TreeNode* root) {
        ans.clear();
        cnt = 0;
        maple.clear();
        dfs(root);
        return ans;
    }
};